p2p/src/peer.rs

```diff
@@ -340,9 +342,13 @@ impl Peer {
 				false
 			}
 			Ok(e) => {
-				*state = State::Disconnected;
-				debug!(LOGGER, "Client {} connection lost: {:?}", self.info.addr, e);
-				self.stop();
+				if State::Disconnected != *state {
+					drop(state);
+					let mut state = self.state.write().unwrap();
+					*state = State::Disconnected;
+					debug!(LOGGER, "Client {} connection lost: {:?}", self.info.addr, e);
+					self.stop();
```

Comment:
Might be worth limiting the scope of the write lock to simply updating the state.
And keeping `stop()` outside the write lock.